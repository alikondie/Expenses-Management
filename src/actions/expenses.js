import { v4 as uuidv4 } from 'uuid';

export const addExpense = ({
  description = '',
  note = '',
  amount = 0,
  createAt = 0
} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuidv4(),
    description,
    note,
    amount,
    createAt
  }
});

export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
