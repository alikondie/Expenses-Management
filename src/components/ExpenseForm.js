import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();
console.log(now.format('D MM YYYY'));

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? props.expense.amount.toString() : '',
      createAt: props.expense ? moment(props.expense.createAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      console.log('setting state');
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createAt) => {
    if (createAt)
      this.setState(() => ({
        createAt,
      }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      const error = 'Please provide description and/or amount';
      this.setState(() => ({
        error,
      }));
    } else {
      this.setState(() => ({
        error: '',
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10),
        createAt: this.state.createAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type='text'
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type='text'
            placeholder='Amount'
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder='Add a note for your expense (optional)'
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
