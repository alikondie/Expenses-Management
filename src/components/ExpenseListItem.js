import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import moment from 'moment';

numeral.register('locale', 'fr', {
  delimiters: {
    thousands: ' ',
    decimal: ',',
  },
  abbreviations: {
    thousand: 'k',
    million: 'm',
    billion: 'b',
    trillion: 't',
  },
  ordinal: function (number) {
    return number === 1 ? 'er' : 'ème';
  },
  currency: {
    symbol: '€',
  },
});

numeral.locale('fr');

export const ExpenseListItem = (props) => (
  <div>
    <Link to={`/edit/${props.id}`}>
      <h3>{props.description}</h3>
    </Link>

    <p>
      {numeral(props.amount).format('0,0[.]00€')} -{' '}
      {moment(props.createAt).format('DD/MM/YYYY')}
    </p>
  </div>
);

export default ExpenseListItem;
