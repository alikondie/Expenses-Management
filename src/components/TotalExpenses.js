import React from 'react';
import { connect } from 'react-redux';
import TotalAmountExpenses from '../selectors/total-expenses';
import selectExpenses from '../selectors/expenses';

export const TotalExpenses = ({ totalExpenses, totalAmount }) => {
  const totalCount = totalExpenses === 1 ? 'expense' : 'expenses';
  return (
    <div>
      <h2>
        Viewing {totalExpenses} {totalCount} with the amount of {totalAmount}
      </h2>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    totalExpenses: visibleExpenses.length,
    totalAmount: TotalAmountExpenses(visibleExpenses),
  };
};

export default connect(mapStateToProps)(TotalExpenses);
