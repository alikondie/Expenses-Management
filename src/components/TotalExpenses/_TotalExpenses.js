import styled from 'styled-components';
import settings from '../../styles/base/settings';
export { Button } from '../../styles/components/Buttons';
export { StyledLink } from '../../styles/components/Link';
export const H1 = styled.h2`
  font-weight: 200;
`;

export const Span = styled.span`
  font-weight: 1500;
`;
