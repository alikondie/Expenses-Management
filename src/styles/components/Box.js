import styled from 'styled-components';
import settings from '../base/settings';
export const Box = styled.div`
  background-color: rgba(
    200,
    200,
    200,
    0.7
  ); /* Used if the image is unavailable */ /* You must set a specified height */
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;

  border-radius: 3px;
  padding: 1rem 1rem 3rem 1rem;
`;

export const Title = styled.h1`
  margin: 0 0 ${settings.m_size} 0;
  line-height: 1;
  color: #111111;
`;

export const Text = styled.p`
  margin: 0 0 ${settings.m_size} 0;
  font-size: 1.3rem;
  word-break: break-all;
  color: #111111;
`;
