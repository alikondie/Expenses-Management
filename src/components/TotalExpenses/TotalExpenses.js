import React from 'react';
import { connect } from 'react-redux';
import TotalAmountExpenses from '../../selectors/total-expenses';
import selectExpenses from '../../selectors/expenses';
import { Container } from '../../styles/components/Container';
import { H1, Span, Button, StyledLink } from './_TotalExpenses';
export const TotalExpenses = ({ totalExpenses, totalAmount }) => {
  const totalCount = totalExpenses === 1 ? 'expense' : 'expenses';
  return (
    <Container>
      <H1>
        Viewing <Span>{totalExpenses}</Span> {totalCount} with the amount of{' '}
        <Span>{totalAmount}</Span>
      </H1>
      <StyledLink to='/create'>
        <Button>Add Expense</Button>
      </StyledLink>
    </Container>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    totalExpenses: visibleExpenses.length,
    totalAmount: TotalAmountExpenses(visibleExpenses),
  };
};

export default connect(mapStateToProps)(TotalExpenses);
