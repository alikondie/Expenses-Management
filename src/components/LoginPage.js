import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';
import { Button } from '../styles/components/LoginPage/Buttons';
import { LoginWrapper } from '../styles/components/LoginPage/LoginWrapper';
import { Box, Title, Text } from '../styles/components/LoginPage/Box';
const LoginPage = ({ startLogin }) => {
  return (
    <LoginWrapper>
      <Box>
        <Title>Expense Management</Title>
        <Text>Manage your expenses with ease</Text>
        <Button onClick={startLogin} id='login'>
          Login with Google
        </Button>
      </Box>
    </LoginWrapper>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
