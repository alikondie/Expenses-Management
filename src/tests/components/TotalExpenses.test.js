import React from 'react';
import { shallow } from 'enzyme';
import { TotalExpenses } from '../../components/TotalExpenses';
import expenses from '../fixtures/expenses';

test('Should render a totalExpenses with 1 data', () => {
  const wrapper = shallow(
    <TotalExpenses totalExpenses={1} totalAmount={0.5} />
  );
  expect(wrapper).toMatchSnapshot();
});

test('Should render a ExpenseListItem with multiple data ', () => {
  const wrapper = shallow(
    <TotalExpenses totalExpenses={45} totalAmount={1354} />
  );
  expect(wrapper).toMatchSnapshot();
});
