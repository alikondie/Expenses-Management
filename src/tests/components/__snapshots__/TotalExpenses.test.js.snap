// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render a ExpenseListItem with multiple data  1`] = `
<div>
  <p>
    Viewing 
    45
     
    expenses
     with the amount of 
    1354
  </p>
  ;
</div>
`;

exports[`Should render a totalExpenses with 1 data 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     with the amount of 
    0.5
  </p>
  ;
</div>
`;
